### ISSUE TITLE
Sprites become corrupted at random as more sprites are spawned

### ISSUE THREAD

**Issue by ProspectPyxis** [bevyengine/bevy#4053]
## Bevy version

0.6.1

## Operating system & version

Pop!_OS 21.10

## What you did

I simply had a system to spawn fireballs from a player position on click. After extensive testing, I've determined that any sort of sprite spawning at all has a chance to cause this, even on a fixed timestep.

## What you expected to happen

Sprites should show up on the screen and work as normal.

## What actually happened

The sprites gradually became corrupted at random as more sprites are spawned. As an example:

![corrupted fireballs](https://user-images.githubusercontent.com/25642180/155878121-9f229ad9-2102-4acc-ace1-509bc5c70a77.png)

The affected sprites seem to be completely random - sometimes the fireballs are corrupted, while other times the player sprite is the one corrupted.

## Additional information

My graphics card information as given by bevy:

```
INFO bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce GTX 960M", vendor: 4318, device: 5019, device_type: DiscreteGpu, backend: Vulkan }
```

I believe this may be an issue specific to my system, though I have no idea how I'd begin debugging it.

**Comment by DJMcNab** [comment]
Can you try running our `bevymark` example? That spawns lots of sprites, and so should hopefully reproduce the issue.

**Comment by ProspectPyxis** [comment]
Huh, strangely the `bevymark` example seems to work with no issues on the v0.6.1 tag - no corrupted sprites. Does this mean I'm doing something potentially incorrectly?

**Comment by mockersf** [comment]
Could you share example code that we could try to reproduce?

**Comment by ProspectPyxis** [comment]
Here's a minimum reproducible example I came up with, the only additional dependency being `alea` for RNG:
```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_startup_system_set(
            SystemSet::new()
                .with_system(setup_camera)
                .with_system(spawn_player),
        )
        .add_system(player_shoot)
        .run();
}

pub fn setup_camera(mut commands: Commands) {
    commands.spawn_bundle(OrthographicCameraBundle::new_2d());
}

pub fn spawn_player(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn_bundle(SpriteBundle {
        texture: asset_server.load("sprites/player.png"),
        ..Default::default()
    });
}

pub fn player_shoot(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mouse_input: Res<Input<MouseButton>>,
) {
    if mouse_input.just_pressed(MouseButton::Left) {
        commands.spawn_bundle(SpriteBundle {
            texture: asset_server.load("sprites/fireball.png"),
            transform: Transform {
                translation: Vec3::new(
                    alea::f32_in_range(-640.0, 640.0),
                    alea::f32_in_range(-480.0, 480.0),
                    0.0,
                ),
                ..Default::default()
            },
            ..Default::default()
        });
    }
}
```

This is part of the resulting window after only a few clicks:
![image](https://user-images.githubusercontent.com/25642180/155885333-5008cbe5-e2bf-4185-8374-e460a20eed03.png)

**Comment by DJMcNab** [comment]
It would probably be helpful to have the sprites you are using too, in case it's something about the image?

**Comment by kirusfg** [comment]
I was able to replicate, this happens around every fifth launch of the game
![2022-02-27-193319_681x374_scrot](https://user-images.githubusercontent.com/32136879/155886703-4524388d-c5bd-4d6b-8149-8517e7aead3b.png)

**Comment by ProspectPyxis** [comment]
I was using these two sprites. The player is from [this pack](https://0x72.itch.io/dungeontileset-ii), and the fireball was custom and exported from GIMP.
![player](https://user-images.githubusercontent.com/25642180/155886841-9cb2112b-2574-4c9a-a20b-5d1403bf68ee.png)
![fireball](https://user-images.githubusercontent.com/25642180/155886845-65258259-6003-4ffa-9dcd-f7f2e068a234.png)

**Comment by cart** [comment]
Just tried reproing on main with some of our pre-packaged sprites and couldn't. Can anyone repro on main with this example?

```rust
use bevy::prelude::*;
use rand::Rng;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_startup_system_set(
            SystemSet::new()
                .with_system(setup_camera)
                .with_system(spawn_player),
        )
        .add_system(player_shoot)
        .run();
}

pub fn setup_camera(mut commands: Commands) {
    commands.spawn_bundle(OrthographicCameraBundle::new_2d());
}

pub fn spawn_player(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn_bundle(SpriteBundle {
        texture: asset_server.load("textures/simplespace/enemy_A.png"),
        ..Default::default()
    });
}

pub fn player_shoot(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mouse_input: Res<Input<MouseButton>>,
) {
    if mouse_input.just_pressed(MouseButton::Left) {
        commands.spawn_bundle(SpriteBundle {
            texture: asset_server.load("textures/simplespace/enemy_B.png"),
            transform: Transform {
                translation: Vec3::new(
                    rand::thread_rng().gen_range(-640.0..640.0),
                    rand::thread_rng().gen_range(-480.0..480.0),
                    0.0,
                ),
                ..Default::default()
            },
            ..Default::default()
        });
    }
}
```

**Comment by cart** [comment]
Can't repro on main with any of the sprites linked above (or by swapping alea back in for the random numbers).

**Comment by cart** [comment]
Can't repro on 0.6.1 either. Nows probably a good time to start sharing relevant environment info:

* Arch Linux: Nvidia GTX 1070 (Proprietary driver version 510.54), Vulkan Backend

**Comment by ProspectPyxis** [comment]
Hm, I couldn't reproduce with the built-in sprites on 0.6.1, but it happens pretty consistently with my own sprites.

I've uploaded a minimal repo that generates the glitch semi-consistently for me [here](https://github.com/ProspectPyxis/bevy-sprite-corruption-example).

As for my environment info:
* Pop!_OS: Nvidia GTX 960M (Proprietary driver version 510.54), Vulkan Backend

**Comment by mockersf** [comment]
I couldn't reproduce with your repo, with 10 runs
`AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: DiscreteGpu, backend: Metal }`

**Comment by cart** [comment]
Yeah sadly I can't repro with your repo either on my machine. I'll try arch+nvidia 750m+proprietary next (on my apple laptop)

**Comment by cart** [comment]
Ok so the good news is ... I got a repro:
![image](https://user-images.githubusercontent.com/2694663/155922872-6edb58dc-8c1e-4b1f-ac20-5cd464a16734.png)

The bad news is ... I got a repro which means this isn't just a problem on your machine :)

Likely somehow related to Linux + nvidia mobile gpus with proprietary drivers?

**Comment by ProspectPyxis** [comment]
Oh thank *goodness* I'm not going crazy.

I'm assuming that's the case too, but in that case this is *way* above my paygrade right now, haha.

**Comment by cart** [comment]
@kirusfg can you share your environment information so we have another datapoint? I'm guessing its linux + a mobile nvidia gpu?

**Comment by cart** [comment]
I can also repro with MaterialMesh2dBundle, which means this isn't related to our sprite batching logic.

**Comment by cart** [comment]
Pulling in @kvark and @cwfitzgerald as a heads up (we'll file a bug once we have a bit more information and wgpu traces ... unless y'all have easy answers).

**Comment by kirusfg** [comment]
Yes, I am on Arch Linux with an Nvidia GeForce 920MX using the proprieatry drivers.

**Comment by cwfitzgerald** [comment]
No particular issue pops out to me, so a semi-minimalized repo would be great

**Comment by kvark** [comment]
At the very least, please file a bug to us and provide a zipped API trace (the smaller the better), please.

**Comment by matidfk** [comment]
Getting the same issue here, running arch linux with proprietary nvidia drivers version 515.65.01
AdapterInfo { name: "NVIDIA GeForce GTX 750 Ti", vendor: 4318, device: 4992, device_type: DiscreteGpu, backend: Vulkan }
![0812_001354](https://user-images.githubusercontent.com/72168331/184259178-21ce2efa-cda2-4a86-8776-bf2471078082.png)

**Comment by mydumpfire** [comment]
I can also reproduce the issue with the repo above. This is the system and adapter logs:

```rust
SystemInfo {
    os: "Linux  Arch Linux",
    kernel: "6.4.3-arch1-1",
    cpu: "Intel(R) Core(TM) i5-4430 CPU @ 3.00GHz",
    core_count: "4",
    memory: "15.6 GiB",
}
```

```rust
AdapterInfo {
    name: "NVIDIA GeForce GT 635",
    vendor: 4318,
    device: 4736,
    device_type: DiscreteGpu,
    driver: "NVIDIA",
    driver_info: "470.199.02",
    backend: Vulkan,
}
```

It would be really nice if this can be fixed. It's a really huge show-stopper.

@kvark
> provide a zipped API trace

Can you tell me how to get the trace?

**Comment by mydumpfire** [comment]
I've managed to make a minimal reproduction repo with the latest 0.11 version of Bevy: https://github.com/mydumpfire/sprite_debug
All it does is load a 16x16 flat blue square png. If the artifacting doesn't occur right away, try to restart the built binary a few times. Personally, the artifact happens around 80-90% of the time.
Zoomed-in screenshot of the artifact: 
![image](https://github.com/bevyengine/bevy/assets/134539278/a037b206-2178-48e8-9a16-6e3532705bde)
 

@kvark I've also made a zipped API trace: [wgpu_trace.zip](https://github.com/bevyengine/bevy/files/12064805/wgpu_trace.zip)

**Comment by zentasumu** [comment]
This appears to match with my report #8355.

I also have an Optimus system.

I found that 64x64 is the minimum (at least on my machine) to prevent such artifacts.
Please test if this holds true for others.

The specific resolution seems to follow the COPY_BYTES_PER_ROW_ALIGNMENT of wgpu, if that is worth mentioning.

I did check that the texture data is not corrupted by transferring it to a bigger texture image, so it seems to be a visual thing.

**Comment by musjj** [comment]
> https://github.com/mydumpfire/sprite_debug

I can reproduce this too.

@zentasumu
You mentioned that a <=2GB vram size might be the culprit, can people in this thread contribute some data on this? I am also personally using a 2GB GPU, so this strengthens the theory.

I also suspected that this is an nvidia driver issue, but:

> I also tried modifying some simple sample vulkan c++ programs to create texture images smaller than 64x64, and I could not reproduce the issue.

So this sounds like a wgpu/bevy bug. Anyone have a clue of the cause of this bug?

**Comment by zentasumu** [comment]
Further down the line, I've had suspicion that this is solely with wgpu.
Using textures with sizes lower than 64x64 have a high chance that it will produce the artifact, but as said before, try the same test on texture related vulkan examples in c/c++ and it should not have an issue.

The randomness of producing the artifact across runs is due to the unordered generation of textures. Plus we also need to consider that Bevy itself generates some of its own, like fallback images.

Backing this, I forgot to say that I've done an additional test of generating checkerboard textures from 1x1 up to 128x128 with only wgpu, and I can confirm that it matches the weird reproduction behavior with multiple textures loaded, in which some are rendered correctly and some don't, excluding 64x64 and up which are always rendered correctly.

Again, this does not appear to be a texture data corruption, which is confirmed by doing an image to image copy, rather the texture sampler is returning wrong colors for some reason.

One thing though, I seem to notice that the artifact changes when using `__NV_PRIME_RENDER_OFFLOAD=1 __VK_LAYER_NV_optimus=NVIDIA_only` vs overriding `VK_ICD_FILENAMES` if that is worth mentioning.

I'm still unsure what wgpu does under the hood which results in the artifact, but following the COPY_BYTES_PER_ROW_ALIGNMENT,  which has a value of 256, seems to work which is where my theory of a minimum 64x64 texture size came from.

**Comment by musjj** [comment]
I've tested the reproduction again with `0.12` and the bug now happens a *lot* more rarely and less harshly. I re-ran the debug build many times and the worst I got was this:

![image](https://github.com/bevyengine/bevy/assets/72612857/2e952792-5fc6-4754-bdd7-e5199a36ea7c)

Can anyone else confirm any experienced differences since `0.12`?
